components/Sataserver/src/sataserver.c

```diff
@@ -0,0 +1,663 @@
+            clients[client].partition_tables[i].num_sectors = phys_partition_tables[part_idx].num_sectors;
+
+            /* Current virtual partition should start right after the last
+             * partition or at the VIRT_START_SECTOR. This will create no holes
+             * in the virtual disk
+             */
+            clients[client].partition_tables[i].start_lba = sectors;
+
+            if ((clients[client].partition_tables[i].start_lba / (MAX_NUM_HEAD * MAX_NUM_SECT)) > MAX_NUM_CYL) {
+                cyl = MAX_NUM_CYL;
+                head = MAX_NUM_HEAD;
+                sec = MAX_NUM_SECT;
+            } else {
+                // convert LBA to CHS
+                cyl = clients[client].partition_tables[i].start_lba / (MAX_NUM_HEAD * MAX_NUM_SECT);
+                head = (clients[client].partition_tables[i].start_lba / MAX_NUM_SECT) % MAX_NUM_HEAD;
+                sec = ((clients[client].partition_tables[i].start_lba % MAX_NUM_SECT) + 1);
+            }
+            clients[client].partition_tables[i].head = head;
+            // The MBR specifies the cylinders as a 10 bit value and the sectors as a 6 bit value
+            // These values are packed in the MBR with the upper 2 bits of the cylinders value
+            // tacked onto the upper 2 bits of the sectors byte
+            clients[client].partition_tables[i].sec_cyl = sec | ((cyl & 0x0300) >> 2) | ((cyl & 0x00FF) << 8);
+
+            end_lba = clients[client].partition_tables[i].start_lba + clients[client].partition_tables[i].num_sectors - 1;
+            if ((end_lba / (MAX_NUM_HEAD * MAX_NUM_SECT)) > MAX_NUM_CYL) {
+                cyl = MAX_NUM_CYL;
+                head = MAX_NUM_HEAD;
+                sec = MAX_NUM_SECT;
+            } else {
+                // convert LBA to CHS
+                cyl = end_lba / (MAX_NUM_HEAD * MAX_NUM_SECT);
+                head = (end_lba / MAX_NUM_SECT) % MAX_NUM_HEAD;
+                sec = ((end_lba % MAX_NUM_SECT) + 1);
+            }
+            clients[client].partition_tables[i].end_head = head;
+            clients[client].partition_tables[i].end_sec_cyl = sec | ((cyl & 0x0300) >> 2) | ((cyl & 0x00FF) << 8);
+
+            /* Increment sectors to figure out the starting LBA for the next partition */
+            sectors += clients[client].partition_tables[i].num_sectors;
+
+            ZF_LOGE("Client(%u): Virtual Partition %i:", clients[client].client_id, i);
+            ZF_LOGE("\tStart: %u, Sectors: %u", clients[client].partition_tables[i].start_lba,
+                    clients[client].partition_tables[i].num_sectors);
+            ZF_LOGE("\tBoot: %u, System ID: 0x%X", clients[client].partition_tables[i].boot,
+                    clients[client].partition_tables[i].sys_id);
+        }
+
+        clients[client].capacity = sectors;
+        ZF_LOGE("\n\t\tCapacity: %llu", clients[client].capacity);
+    }
+
+    ZF_LOGE("Finished all initialisation");
```

Comment:
These log levels should be Info level too

Reply:
Yep.